guide/src/class/protocols.md

```diff
@@ -21,22 +21,40 @@ The magic methods handled by PyO3 are very similar to the standard Python ones o
  - Magic methods for the sequence protocol
 
 When PyO3 handles a magic method, a couple of changes apply compared to other `#[pymethods]`:
-  - The `#[pyo3(text_signature = "...")]` attribute is not allowed
-  - The types of the arguments are fixed according to the magic method
+ - The `#[pyo3(text_signature = "...")]` attribute is not allowed
+ - The signature is restricted to match the magic method
+
+The following sections list of all magic methods PyO3 currently handles.  The
+given signatures should be interpreted as follows:
+ - All methods take a receiver as first argument, which can be `&self`, `&mut
+   self` or a `PyCell` reference like `self_: PyRef<Self>` and `self_:
+   PyRefMut<Self>`, as described [here](../class.md#inheritance).
+ - An optional `Python<'py>` argument is always allowed as the first argument.
+ - Return values can be optionally wrapped in `PyResult`.
+ - `object` means that any type is allowed that can be extracted from a Python
+   object (if argument) or converted to a Python object (if return value).
+ - Other types must match what's given, e.g. `pyo3::basic::CompareOp` for
+   `__richcmp__`'s second argument.
+ - For the comparison and arithmetic methods, extraction errors are not
+   propagated as exceptions, but lead to a return of `NotImplemented`.
+ - For some slots, the return values are not restricted by PyO3, but checked by
+   the Python interpreter. For example, `__str__` needs to return a string
+   object.  This is indicated by `object (Python type)`.
 
-The following sections list of all magic methods PyO3 currently handles:
 
 #### Basic object customization
 
-  - `__str__`
-  - `__repr__`
-  - `__hash__`
-  - `__richcmp__`
-  - `__getattr__`
-  - `__setattr__`
-  - `__delattr__`
-  - `__bool__`
-  - `__call__`
+  - `__str__() -> object (str)`
```

Comment:
Although arguably redundant, should receivers be included?

```suggestion
  - `__str__(&self) -> object (str)`
```

I guess we don't want to pin users to a single receiver type, so seeing this here will either remind them to include it or confuse them!

Reply:
Or use something like `<self>`.

Reply:
`<self>` sounds like a great idea to me; seems like a good placeholder that's clearly not valid syntax.